flow/connectors/mongo/mongo.go

```diff
@@ -0,0 +1,431 @@
+	"github.com/PeerDB-io/peerdb/flow/shared"
+	"github.com/PeerDB-io/peerdb/flow/shared/types"
+)
+
+const (
+	DefaultDocumentKeyColumnName  = "_id"
+	DefaultFullDocumentColumnName = "_full_document"
+)
+
+type MongoConnector struct {
+	*metadataStore.PostgresMetadata
+	config *protos.MongoConfig
+	client *mongo.Client
+	logger log.Logger
+}
+
+func NewMongoConnector(ctx context.Context, config *protos.MongoConfig) (*MongoConnector, error) {
+	logger := internal.LoggerFromCtx(ctx)
+	pgMetadata, err := metadataStore.NewPostgresMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := mongo.Connect(options.Client().
+		SetAppName("PeerDB Mongo Connector").
+		SetReadPreference(readpref.Primary()).
+		SetCompressors([]string{"zstd", "snappy"}).
+		ApplyURI(config.Uri))
+	if err != nil {
+		return nil, err
+	}
+	return &MongoConnector{
+		PostgresMetadata: pgMetadata,
+		config:           config,
+		client:           client,
+		logger:           logger,
+	}, nil
+}
+
+func (c *MongoConnector) Close() error {
+	if c != nil && c.client != nil {
+		// Use a timeout to ensure the disconnect operation does not hang indefinitely
+		timeout, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel()
+		return c.client.Disconnect(timeout)
+	}
+	return nil
+}
+
+func (c *MongoConnector) ConnectionActive(ctx context.Context) error {
+	// Use a ping command to check if the connection is active
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	if err := c.client.Ping(ctx, nil); err != nil {
+		return fmt.Errorf("failed to ping MongoDB: %w", err)
+	}
+	return nil
+}
+
+func (c *MongoConnector) GetVersion(ctx context.Context) (string, error) {
+	db := c.client.Database("admin")
+
+	var buildInfoDoc bson.M
+	if err := db.RunCommand(ctx, bson.D{bson.E{Key: "buildInfo", Value: 1}}).Decode(&buildInfoDoc); err != nil {
+		return "", fmt.Errorf("failed to run buildInfo command: %w", err)
+	}
+
+	version, ok := buildInfoDoc["version"].(string)
+	if !ok {
+		return "", fmt.Errorf("buildInfo.version is not a string, but %T", buildInfoDoc["version"])
+	}
+	return version, nil
+}
+
+func (c *MongoConnector) ValidateMirrorSource(ctx context.Context, cfg *protos.FlowConnectionConfigs) error {
+	if cfg.DoInitialSnapshot && cfg.InitialSnapshotOnly {
+		return nil
+	}
+
+	// Check if MongoDB is configured as a replica set
+	var result bson.M
+	if err := c.client.Database("admin").RunCommand(ctx, bson.D{
+		{Key: "replSetGetStatus", Value: 1},
+	}).Decode(&result); err != nil {
+		return fmt.Errorf("failed to get replica set status: %w", err)
+	}
+
+	// Check if this is a replica set
+	if _, ok := result["set"]; !ok {
+		return errors.New("MongoDB is not configured as a replica set, which is required for CDC")
+	}
+	if myState, ok := result["myState"].(int32); !ok || myState != 1 {
+		return fmt.Errorf("MongoDB is not the primary node in the replica set, current state: %d", myState)
```

Comment:
myState will be 0 here when it failed to cast to int32. Should have !ok error separate from branch using myState